File: stmt.cpp
#include "stmt.hpp"
#include "environment.hpp"
#include "expr.hpp"
#include "token.hpp"


// Stmt classes
auto Stmt::delete_object(Stmt* object) -> void {
    delete object;
}


// PrintStmt
auto PrintStmt::create_object(Expr* expr) -> PrintStmt* {
    return new PrintStmt { expr };
}

PrintStmt::PrintStmt(Expr* expr) :
    expr { expr }
    {
    }

PrintStmt::~PrintStmt() {
    Expr::delete_object(expr);
}

auto PrintStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_printstmt(this);
}

auto PrintStmt::type() -> StmtType {
    return StmtType::k_print;
}


// PrintlnStmt
auto PrintlnStmt::create_object(Expr* expr) -> PrintlnStmt* {
    return new PrintlnStmt { expr };
}

PrintlnStmt::PrintlnStmt(Expr* expr) :
    expr { expr }
    {
    }

PrintlnStmt::~PrintlnStmt() {
    Expr::delete_object(expr);
}

auto PrintlnStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_printlnstmt(this);
}

auto PrintlnStmt::type() -> StmtType {
    return StmtType::k_print;
}


// ExpressionStmt
auto ExpressionStmt::create_object(Expr* expr) -> ExpressionStmt* {
    return new ExpressionStmt { expr };
}

ExpressionStmt::ExpressionStmt(Expr* expr) :
    expr { expr }
    {
    }

ExpressionStmt::~ExpressionStmt() {
    Expr::delete_object(expr);
}

auto ExpressionStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_expressionstmt(this);
}

auto ExpressionStmt::type() -> StmtType {
    return StmtType::k_expression;
}


// VarStmt
auto VarStmt::create_object(Token* identifier, Expr* expr) -> VarStmt* {
    return new VarStmt { identifier, expr };
}

VarStmt::VarStmt(Token* identifier, Expr* expr) :
    identifier { Token::create_object(*identifier) },
    expr { expr }
    {
    }

VarStmt::~VarStmt() {
    Token::delete_object(identifier);
    Expr::delete_object(expr);
}

auto VarStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_varstmt(this);
}

auto VarStmt::type() -> StmtType {
    return StmtType::k_var;
}


// ConstStmt
auto ConstStmt::create_object(Token* identifier, Expr* expr) -> ConstStmt* {
    return new ConstStmt { identifier, expr };
}

ConstStmt::ConstStmt(Token* identifier, Expr* expr) :
    identifier { Token::create_object(*identifier) },
    expr { expr }
    {
    }

ConstStmt::~ConstStmt() {
    Token::delete_object(identifier);
    Expr::delete_object(expr);
}

auto ConstStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_conststmt(this);
}

auto ConstStmt::type() -> StmtType {
    return StmtType::k_const;
}


// BlockStmt
auto BlockStmt::create_object(std::vector<Stmt*>& statements) -> BlockStmt* {
    return new BlockStmt { statements };
}

BlockStmt::BlockStmt(std::vector<Stmt*>& statements) :
    statements { statements }
    {
    }

BlockStmt::~BlockStmt() {
    for (Stmt* stmt :statements) {
        Stmt::delete_object(stmt);
    }
}

auto BlockStmt::accept(StmtVisitor* visitor) -> void {
    return const_cast<StmtVisitor*>(visitor)->visit_blockstmt(this, Environment::create_object());
}

auto BlockStmt::type() -> StmtType {
    return StmtType::k_block;
}


// IfStmt
IfStmt::IfStmt(std::vector<std::pair<Expr*, Stmt*>>& statements) :
    statements { statements }
    {
    }

IfStmt::~IfStmt() {
    for (auto item : statements) {
        Expr::delete_object(item.first);
        Stmt::delete_object(item.second);
    }
}

auto IfStmt::create_object(std::vector<std::pair<Expr*, Stmt*>>& statements) -> IfStmt* {
    return new IfStmt { statements };
}

auto IfStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_ifstmt(this);
}

auto IfStmt::type() -> StmtType {
    return StmtType::k_if;
}


// ForStmt
ForStmt::ForStmt(Expr* condition, Stmt* block_stmt) :
    condition { condition },
    block_stmt { block_stmt }
    {
    }

ForStmt::~ForStmt() {
    Expr::delete_object(condition);
    Stmt::delete_object(block_stmt);
}

auto ForStmt::create_object(Expr* condition, Stmt* block_stmt) -> ForStmt* {
    return new ForStmt { condition, block_stmt };
}

auto ForStmt::accept(StmtVisitor* visitor) -> void {
    return const_cast<StmtVisitor*>(visitor)->visit_forstmt(this);
}

auto ForStmt::type() -> StmtType {
    return StmtType::k_for;
}


// BreakStmt
BreakStmt::BreakStmt(uint32_t line) :
    line { line }
    {
    }

auto BreakStmt::create_object(uint32_t line) -> BreakStmt* {
    return new BreakStmt { line };
}

auto BreakStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_breakstmt(this);
}

auto BreakStmt::type() -> StmtType {
    return StmtType::k_break;
}


// ContinueStmt
ContinueStmt::ContinueStmt(uint32_t line, Expr* step_expr) :
    line { line },
    step_expr { step_expr }
    {
    }

auto ContinueStmt::create_object(uint32_t line, Expr* step_expr) -> ContinueStmt* {
    return new ContinueStmt { line, step_expr };
}

auto ContinueStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_continuestmt(this);
}

auto ContinueStmt::type() -> StmtType {
    return StmtType::k_continue;
}


// ReturnStmt
ReturnStmt::ReturnStmt(Token* token, Expr* value) :
    token { token },
    value { value }
    {
    }

ReturnStmt::~ReturnStmt() {
    Expr::delete_object(value);
}

auto ReturnStmt::create_object(Token* token, Expr* value) -> ReturnStmt* {
    return new ReturnStmt { token, value };
}

auto ReturnStmt::accept(StmtVisitor* visitor) -> void {
    return visitor->visit_returnstmt(this);
}

auto ReturnStmt::type() -> StmtType {
    return StmtType::k_return;
}

File: lib/stmt.ts
import IExpr, { ClassExpr, FuncExpr } from "./expr"
import Token from "./token"

export default interface IStmt {
    accept(visitor: IStmtVisitor): void
}

export class PrintStmt implements IStmt {
    private constructor(private _expr: IExpr) {}

    static createInstance(expr: IExpr) {
        return new PrintStmt(expr)
    }

    expr() {
        return this._expr
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitPrintStmt(this)
    }
}

export class PrintlnStmt implements IStmt {
    private constructor(private _expr: IExpr) {}

    static createInstance(expr: IExpr) {
        return new PrintlnStmt(expr)
    }

    expr() {
        return this._expr
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitPrintlnStmt(this)
    }
}

export class LetStmt implements IStmt {
    private constructor(private _identifier: Token, private _expr: IExpr) {}

    static createInstance(identifier: Token, expr: IExpr) {
        return new LetStmt(identifier, expr)
    }

    identifier() {
        return this._identifier
    }

    expr() {
        return this._expr
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitLetStmt(this)
    }
}

export class ConstStmt implements IStmt {
    private constructor(private _identifier: Token, private _expr: IExpr) {}

    static createInstance(identifier: Token, expr: IExpr) {
        return new ConstStmt(identifier, expr)
    }

    identifier() {
        return this._identifier
    }

    expr() {
        return this._expr
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitConstStmt(this)
    }
}

export class BlockStmt implements IStmt {
    private constructor(private _stmts: IStmt[]) {}

    static createInstance(stmts: IStmt[]) {
        return new BlockStmt(stmts)
    }

    stmts() {
        return this._stmts
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitBlockStmt(this)
    }
}

export class IfStmt implements IStmt {
    private constructor(private _stmtTable: [IExpr, BlockStmt][]) {}

    static createInstance(stmtTable: [IExpr, BlockStmt][]) {
        return new IfStmt(stmtTable)
    }

    stmtTable() {
        return this._stmtTable
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitIfStmt(this)
    }
}

export class ForWrapperStmt implements IStmt {
    private constructor(private _forStmt: IStmt) {}

    static createInstance(forStmt: IStmt) {
        return new ForWrapperStmt(forStmt)
    }

    forStmt() {
        return this._forStmt
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitForWrapperStmt(this)
    }
}

export class InfiniteForStmt implements IStmt {
    private constructor(private _stmts: BlockStmt) {}

    static createInstance(stmts: BlockStmt) {
        return new InfiniteForStmt(stmts)
    }

    stmts() {
        return this._stmts
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitInfiniteForStmt(this)
    }
}

export class WhileForStmt implements IStmt {
    private constructor(private _condition: IExpr, private _stmts: BlockStmt) {}

    static createInstance(condition: IExpr, stmts: BlockStmt) {
        return new WhileForStmt(condition, stmts)
    }

    condition() {
        return this._condition
    }

    stmts() {
        return this._stmts
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitWhileForStmt(this)
    }
}

export class ForStmt implements IStmt {
    private constructor(
        private _init: IExpr,
        private _condition: IExpr,
        private _step: IExpr,
        private _stmts: BlockStmt
    ) {}

    static createInstance(
        init: IExpr,
        condition: IExpr,
        step: IExpr,
        stmts: BlockStmt
    ) {
        return new ForStmt(init, condition, step, stmts)
    }

    init() {
        return this._init
    }

    condition() {
        return this._condition
    }

    step() {
        return this._step
    }

    stmts() {
        return this._stmts
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitForStmt(this)
    }
}

export class LetForStmt implements IStmt {
    private constructor(
        private _init: LetStmt,
        private _condition: IExpr,
        private _step: IExpr,
        private _stmts: BlockStmt
    ) {}

    static createInstance(
        init: LetStmt,
        condition: IExpr,
        step: IExpr,
        stmts: BlockStmt
    ) {
        return new LetForStmt(init, condition, step, stmts)
    }

    init() {
        return this._init
    }

    condition() {
        return this._condition
    }

    step() {
        return this._step
    }

    stmts() {
        return this._stmts
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitLetForStmt(this)
    }
}

export class BreakStmt implements IStmt {
    private constructor(private _keyword: Token) {}

    static createInstance(keyword: Token) {
        return new BreakStmt(keyword)
    }

    keyword() {
        return this._keyword
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitBreakStmt(this)
    }
}

export class ContinueStmt implements IStmt {
    private constructor(private _keyword: Token) {}

    static createInstance(keyword: Token) {
        return new ContinueStmt(keyword)
    }

    keyword() {
        return this._keyword
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitContinueStmt(this)
    }
}

export class FuncStmt implements IStmt {
    private constructor(private _name: Token, private _expr: FuncExpr) {}

    static createInstance(name: Token, expr: FuncExpr) {
        return new FuncStmt(name, expr)
    }

    name() {
        return this._name
    }

    expr() {
        return this._expr
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitFuncStmt(this)
    }
}

export class ReturnStmt implements IStmt {
    private constructor(private _keyword: Token, private _value: IExpr) {}

    static createInstance(keyword: Token, value: IExpr) {
        return new ReturnStmt(keyword, value)
    }

    keyword() {
        return this._keyword
    }

    value() {
        return this._value
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitReturnStmt(this)
    }
}

export class ClassStmt implements IStmt {
    private constructor(private _name: Token, private _expr: ClassExpr) {}

    static createInstance(name: Token, expr: ClassExpr) {
        return new ClassStmt(name, expr)
    }

    name() {
        return this._name
    }

    expr() {
        return this._expr
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitClassStmt(this)
    }
}

export class ExprStmt implements IStmt {
    private constructor(private _expr: IExpr) {}

    static createInstance(expr: IExpr) {
        return new ExprStmt(expr)
    }

    expr() {
        return this._expr
    }

    accept(visitor: IStmtVisitor): void {
        visitor.visitExprStmt(this)
    }
}

export interface IStmtVisitor {
    execute(stmts: IStmt[]): void

    visitPrintStmt(stmt: PrintStmt): void
    visitPrintlnStmt(stmt: PrintlnStmt): void
    visitLetStmt(stmt: LetStmt): void
    visitConstStmt(stmt: ConstStmt): void
    visitBlockStmt(stmt: BlockStmt): void
    visitIfStmt(stmt: IfStmt): void
    visitForWrapperStmt(stmt: ForWrapperStmt): void
    visitInfiniteForStmt(stmt: InfiniteForStmt): void
    visitWhileForStmt(stmt: WhileForStmt): void
    visitForStmt(stmt: ForStmt): void
    visitLetForStmt(stmt: LetForStmt): void
    visitBreakStmt(stmt: BreakStmt): void
    visitContinueStmt(stmt: ContinueStmt): void
    visitFuncStmt(stmt: FuncStmt): void
    visitReturnStmt(stmt: ReturnStmt): void
    visitClassStmt(stmt: ClassStmt): void
    visitExprStmt(stmt: ExprStmt): void
}
